2018/python/file_generator.py

```diff
@@ -487,7 +487,115 @@ def edit_report(self):
+		# the dictionary.
+		for key, value in dictionary.items():
+			lar_df[key] = value
+
+		checker = rules_engine(lar_schema=self.lar_schema_df, 
+					ts_schema=self.ts_schema_df, 
+					crosswalk_data=self.crosswalk_data)
+
+		#Produces a report as to which syntax or validity
+		#edits have passed or failed based on logic in the rules_engine.
+		#Loads the TS and LAR dataframes into the checker object. 
+
+		checker.load_data_frames(ts_df, lar_df)
+
+		for func in dir(checker):
+			if func[:1] in ("s", "v") and func[1:4].isdigit()==True:
+				getattr(checker, func)()
+
+	
+		#Produces a report as to which syntax or validity
+		#edits have passed or failed based on logic in the rules_engine.
+		for func in dir(checker):
+			if func[:1] in ("s", "v") and func[1:4].isdigit()==True:
+				getattr(checker, func)()
+		
+		#Creates a results dataframe and keeps the results that 
+		#have failed. 
+		res_df = pd.DataFrame(checker.results)
+		new_df = res_df[(res_df['status']=='failed')]
+
+		# The function ignores TS edits and drops results related
+		# to edit fails from the TS.  
+		new_df = new_df[new_df['row_ids'] != 'TS']
+
+		if len(new_df) == 0:
+			#If there are no syntax or validity edits
+			#the data is written to a new directory for quality 
+			#test files that pass syntax and validity edits. 
+
+			utils.write_file(path=self.filepaths['quality_pass_s_v_filepath'].format(
+				bank_name=self.data_map['name']['value']), 
+				ts_input=ts_df, 
+				lar_input=lar_df, 
+				name=quality_filename)
+
+		#The case if there are rows that failed syntax or validity edits.
+		
+		else: 
+			#Creates an empty list for storing row numbers
+			#where edits have failed. 
+			uli_list = []
+
+			#Iterates through each row and appends the list of ULI's
+			#of rows where syntax or validity edits have failed. 
+			for index, row in new_df.iterrows():
+				uli_list.append(row.row_ids)
+
+			#Drops not-a-numbers from the ULI list. 
+			if np.nan in uli_list:
+				uli_list.remove(np.nan)
+
+			#Creates a new list to store the ULI's without nested brackets. 
+			new_list = []
+			for i in range(len(uli_list)):
+				for n in range(len(uli_list[i])):
+					new_list.append(uli_list[i][n])
+
+			#Creates a list that removes ULI's that are repeated. 
+			unique_list = []
+			for i in new_list:
+				if i not in unique_list:
+					unique_list.append(i)
+
+			#Creates a list of row numbers corresponding to the
+			#ULI's that have failed syntax or validity edits. 
+			bad_rows = []
+			for index, row in lar_df.iterrows():
+				if row['uli'] in unique_list:
+					failed_uli = row['uli']
+					bad_rows.append(lar_df[lar_df['uli']==failed_uli].index.values.astype(int)[0])
+
+			#Drops all rows that failed syntax or validity edits
+			#from the original LAR dataframe. 
+			lar_df = lar_df.drop(bad_rows)
+
+			#Takes the first row of data. 
+			lar_row = lar_df[0:1]
+
+			print('Length of data to be added: ' + str(len(lar_row.index)))
```

Comment:
Can you explain the why of what this is doing?

Reply:
This is checking that the function is producing a row. The length of data being added is supposed to be 1.

Reply:
I think I'm confused about adding a single row. to what is the row being added?

The code above does a lot of work to get a list of bad ULIs, then drop them from a dataframe. Is there a reason to not use the entire dataframe of remaining good rows?

Reply:
The purpose is to make a row that tests a particular condition, and only one row would be needed. The function outputs only on LAR row (I am going to change the language around data being added).